Add PA sync service

gzh-terry opened the pull request.

huangyulong3 reviewed (approved):
LGTM

huangyulong3 reviewed (approved):
LGTM

huangyulong3 reviewed (approved):
LGTM

huangyulong3 reviewed (approved):
LGTM

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 commented on framework/include/bt_pa_sync.h:
```diff
@@ -0,0 +1,63 @@
+
+    /** Present if available: Broadcast_ID from Broadcast Audio Announcement.
+     *  Otherwise, `BT_INVALID_BROADCAST_ID` */
+    uint32_t broadcast_id;
```
Change it to `aid` will be better.

gzh-terry replied:
The official name is `Broadcast_ID`, not a "Advertiser ID"
<img width="942" height="596" alt="image" src="https://github.com/user-attachments/assets/faaa4f11-4eae-48d5-896c-a7c950387d9c" />
From Basic Audio Profle v1.0.1, Section 3.7.2.1

huangyulong3 commented on framework/include/bt_pa_sync.h:
```diff
@@ -0,0 +1,63 @@
+    uint32_t broadcast_id;
+
+    /** UTF-8 string of the `Broadcast_Name` field in `AdvData` (if present) */
+    char broadcast_name[BT_BROADCAST_NAME_MAX_LEN + 1];
```
Change it to `adv_name` will be more easy to understand.

gzh-terry replied:
This name cames from Assigned Numbers, Section 6.12.6.13
<img width="673" height="257" alt="image" src="https://github.com/user-attachments/assets/f83ae539-4cfb-4397-928f-697bd9c9e4d0" />

huangyulong3 commented on framework/include/bt_pa_sync.h:
```diff
@@ -0,0 +1,63 @@
+
+    /** UTF-8 string of the `Broadcast_Name` field in `AdvData` (if present) */
+    char broadcast_name[BT_BROADCAST_NAME_MAX_LEN + 1];
+} bt_pa_sync_info_t;
```
Why there is no `phy`, `interval` and `num_subevents`, `subevent_interval`, `response_slot_delay`, `response_slot_spacing` when sync rsp is enable.

gzh-terry replied:
These info is parsed from secondary advertising. Especially from "LE Extended Advertising Report event".

- `phy`, yes, we have `phy` for primary and secondary channels, I can add them later.
- `Interval`, or `Periodic Advertising Interval`. is included in `ble_scan_result_t`now (also in this PR). So the upper layer can get the interval before extracting AdvData.
- `num_subevents`and the other parameters cames from "LE Periodic Advertising Sync Established event", which is not included in this PR.

huangyulong3 commented on framework/include/bluetooth.h:
```diff
@@ -349,6 +349,10 @@ typedef uint8_t bt_128key_t[16];
 #define BT_UUID_MAX_NUM (32)
 #define BT_UUID_128_LEN (16)
 
+#define BT_INVALID_BROADCAST_ID (0xFFFF)
```
Change `BROADCAST_ID` to AID will be better.

gzh-terry replied:
Same, I would prefer to keep the official name without further reasons.

huangyulong3 commented on framework/include/bluetooth.h:
```diff
@@ -349,6 +349,10 @@ typedef uint8_t bt_128key_t[16];
 
+#define BT_INVALID_BROADCAST_ID (0xFFFF)
+
+#define BT_BROADCAST_NAME_MAX_LEN (128)
```
Change BROADCAST to ADV will be better.

gzh-terry replied:
For the same reason, I would prefer to keep the official names.

huangyulong3 reviewed (commented):
LGTM.

huangyulong3 commented on tools/bis_sink.c:
```diff
@@ -37,18 +40,87 @@ static void usage(void)
+static void on_scan_status(bt_scanner_t* scanner, uint8_t status)
+{
+    PRINT("%s, status = %d", __func__, status);
+    if (g_bis_sink->scanner != scanner)
```
Add some log will more easy to debug.

gzh-terry replied:
Done

huangyulong3 commented on tools/auracast_sink.c:
```diff
@@ -37,18 +40,87 @@ static void usage(void)
+        return;
+
+    if (status != BT_STATUS_SUCCESS)
+        g_bis_sink->scanner = NULL;
```
Same as above.

gzh-terry replied:
`status` already printed.

huangyulong3 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +213,26 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+    uint16_t offset = 0;
+
+    if (!cb)
+        false;
```
`return false`?

gzh-terry replied:
Lol my fault.

huangyulong3 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +213,26 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+        false;
+
+    while (offset < len) {
+        ad = (adv_data_t*)&data[offset];
```
`ad = (adv_data_t*)&data[offset]; `This line may cause memory alignment issues.

Here you are casting data[offset] to the `adv_data_t*` type. Assuming data is just a raw byte array, the adv_data_t structure may have memory alignment problems on certain platforms.

gzh-terry replied:
Yes, I noticed it as well since there were warnings when printing the data.
However, I’d like to follow the existing code patterns, like `advertiser_data_dump`.
Otherwise, this PR would include many changes irrelevant to Auracast itself.

huangyulong3 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +213,26 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+
+    while (offset < len) {
+        ad = (adv_data_t*)&data[offset];
+        if (ad->len == 0)
```
If `ad->len == 0`, you should indeed skip the current advertisement data. However, the current approach directly returns false, which means the entire parsing process terminates—potentially not meeting practical requirements. You might want to proceed to the next advertisement or at least print a warning. We recommend modifying it to:
```
if (ad->len == 0) {
    offset += sizeof(ad->len);  // Skip this entry
    continue;  // Go to next item
}
```

huangyulong3 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +213,26 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+        if (offset > len)
+            return false; /**< Incomplete AD Data */
+
+        cb(ad, context);
```
make sure the `cb` isn't NULL.

gzh-terry replied:
Checked in advance.

huangyulong3 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -85,41 +85,96 @@ static bt_status_t sal_send_req(sal_scan_req_t* req)
-    le_eir_data_t* eir = user_data;
+    le_adv_data_t* adv_data = user_data;
+
+    adv_data->value[adv_data->length++] = data->data_len + 1;
```
user_data == NULL → Direct dereference crash
data == NULL → Accessing data->data_len crash
Change to:
```
if (!data || !user_data) {
    return false;
}
```

gzh-terry replied:
Hi @huangyulong3 

`device_found` and `bt_data_parse` are useless at SAL and are removed now.

huangyulong3 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -85,41 +85,96 @@ static bt_status_t sal_send_req(sal_scan_req_t* req)
-    le_eir_data_t* eir = user_data;
+    le_adv_data_t* adv_data = user_data;
+
+    adv_data->value[adv_data->length++] = data->data_len + 1;
```
`data->data_len + 1` may overflow
```
if (data->data_len >= 0xFF) {
    return false;
}
```

gzh-terry replied:
Hi @huangyulong3 

`device_found` and `bt_data_parse` are useless at SAL and are removed now.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (dismissed):
LGTM

zhangyuan376 commented on tools/bis_sink.c:
```diff
@@ -37,18 +40,87 @@ static void usage(void)
+
+    status = bt_pa_sync_parse_adv_data(info, result);
+    if (status != BT_STATUS_SUCCESS)
+        return;
```
info should be freed

gzh-terry replied:
It's freed later on `exit:`

zhangyuan376 commented on tools/auracast_sink.c:
```diff
@@ -37,18 +40,87 @@ static void usage(void)
 
 void bis_sink_command_uninit(void* handle)
 {
+    thread_loop_exit(&g_bis_sink->loop);
```
Where is the loop started?

gzh-terry replied:
Not in this PR.

zhangyuan376 commented on framework/socket/bt_pa_sync.c:
```diff
@@ -0,0 +1,31 @@
+
+static void adv_data_parsed(adv_data_t* data, void* context);
+
+bt_status_t bt_pa_sync_parse_adv_data(bt_pa_sync_info_t* info, const ble_scan_result_t* result)
```
Why is client-to-server communication not needed, but still placed in socket?

gzh-terry replied:
Will move to /common later.

gzh-terry replied:
Done

liuX10 commented on service/src/pa_sync_event.h:
```diff
@@ -0,0 +1,74 @@
+#define PA_SYNC_EVENT_DATA_FAILED (0xFF) /**< Failed to receive an AUX_SYNC_SUBEVENT_IND PDU */
+
+typedef enum {
+    SYNC_ESTABLISHED,
```
Do these three types correspond to the three per_adv HCI events?
If so, would replacing `SYNC_TERMINATED` with `SYNC_LOST` be more appropriate?

gzh-terry replied:
This is not necessary a sync lost.
Probably terminate by our self.

gzh-terry replied:
> Do these three types correspond to the three per_adv HCI events? If so, would replacing `SYNC_TERMINATED` with `SYNC_LOST` be more appropriate?

No, this is not a direct translation of HCI events.

liuX10 commented on service/src/pa_sync_service.c:
```diff
@@ -17,24 +17,51 @@
 #include "utils/log.h"
 
+typedef struct pa_sync_info {
+    int nothing_but_wait_for_the_first_variable;
```
The variable name is too long.

gzh-terry replied:
Would rename it later in the next PR until there is a variable.

liuX10 commented on framework/include/bluetooth.h:
```diff
@@ -349,6 +349,10 @@ typedef uint8_t bt_128key_t[16];
 
+#define BT_INVALID_BROADCAST_ID (0xFFFF)
+
+#define BT_BROADCAST_NAME_MAX_LEN (128)
```
The `broadcast_name` field has a maximum length of 128 Bytes, which should include the terminating null character ‘\0’.
so 127 maybe better?

gzh-terry replied:
No it doesn't, because there is actually a `Length` field for this `Broadcast_Name`.
Note that any strings in AD data are not necessarily (and usually not) end with a '\0'

zhongzhijie1 commented on framework/include/bt_le_scan.h:
```diff
@@ -91,11 +91,14 @@ typedef enum {
+    uint8_t length; /* length of `adv_data` */
+    uint16_t interval; /* periodic advertising interval in 1.25 microseconds, 0 if not presented */
+    uint8_t sid; /* advertising set identifier, valid from 0x00 to 0x0F, 0xFF if not provided */
     uint8_t pad[1];
```
15 + 1 = 16， read adv_data start from 16

gzh-terry replied:
> 15 + 1 = 16， read adv_data start from 16

Lol, AI not so smart.

zhongzhijie1 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -39,9 +39,9 @@ typedef struct {
+    uint8_t value[229]; /**< 229 for extended advertising report */
     uint8_t length;
-} le_eir_data_t;
+} le_adv_data_t;
```
lgtm

zhongzhijie1 commented on tools/bis_sink.c:
```diff
@@ -59,20 +69,44 @@ static void on_scan_result(bt_scanner_t* scanner, ble_scan_result_t* result)
+        _STRCAT(log, ", txpower:%d", result->tx_power);
+
+    if (result->rssi != 0x7F)
+        _STRCAT(log, ", rssi:%d", result->rssi);
```
it seems _STRCAT can be utils func

gzh-terry replied:
Probably in the future, we have many functions that can be moved to `bt_utils.h`

gzh-terry commented on framework/include/bt_pa_sync.h:
```diff
@@ -0,0 +1,60 @@
+ * @param[in] result Advertising report received from `on_scan_result_cb_t`
+ *
+ * @return `BT_STATUS_SUCCESS` if a SyncInfo field is included
+ * @return `BT_STATUS_NOT_FOUND` if no SyncInfo field is found
```
Done.

gzh-terry commented on framework/include/bluetooth.h:
```diff
@@ -349,6 +349,10 @@ typedef uint8_t bt_128key_t[16];
 #define BT_UUID_MAX_NUM (32)
 #define BT_UUID_128_LEN (16)
 
+#define BT_INVALID_BROADCAST_ID (0xFFFF)
```
Done.

gzh-terry commented on service/stacks/include/sal_pa_sync_interface.h:
```diff
@@ -0,0 +1,153 @@
+     *  reception. Range from 0x0000 to 0x01F3 */
+    uint16_t skip;
+
+    /** Synchronization timeout for the periodic advertising train, measured in 10 microseconds.
```
Done

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+void bis_sink_command_uninit(void* handle)
+{
```
Corrected later in the next PR.

gzh-terry commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +217,91 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+    memset(info, 0x00, sizeof(bt_pa_sync_info_t));
+    if (!advertiser_data_parse(result->adv_data, result->length, adv_data_parsed, info))
+        return BT_STATUS_FAIL;
+
```
Done

gzh-terry commented on CMakeLists.txt:
```diff
@@ -180,6 +180,11 @@ if(CONFIG_BLUETOOTH)
 
+    if(CONFIG_BLUETOOTH_PA_SYNC)
+      list(APPEND CSRCS ${BLUETOOTH_DIR}/framework/socket/bt_pa_sync.c)
+      list(APPEND CSRCS ${BLUETOOTH_DIR}/service/ipc/socket/src/bt_socket_pa_sync.c)
```
Used in the next PR

gzh-terry commented on Makefile:
```diff
@@ -150,6 +150,11 @@ CSRCS += framework/socket/bt_le_scan.c
 
+ifeq ($(CONFIG_BLUETOOTH_PA_SYNC), y)
+CSRCS += framework/socket/bt_pa_sync.c
+CSRCS += service/ipc/socket/src/bt_socket_pa_sync.c
```
Used in the next PR

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,207 @@
+
+void bis_sink_command_uninit(void* handle)
+{
+    thread_loop_exit(&g_bis_sink->loop);
```
Used in the next PR

gzh-terry commented on tools/bis_sink.c:
```diff
@@ -0,0 +1,207 @@
+#define _STRCAT(dst, src, ...)                                         \
+    do {                                                               \
+        size_t _len = strlen(dst);                                     \
+        if (_len + strlen(src) > _LOG_STR_LEN - 1)                     \
```
Done.

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,207 @@
+    if (!g_bis_sink->scanner) {
+        PRINT("Not scanning");
+        return CMD_USAGE_FAULT;
+    }
```
Done

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,207 @@
+    thread_loop_exit(&g_bis_sink->loop);
+    free(g_bis_sink);
+    g_bis_sink = NULL;
+}
```
Done.

gzh-terry commented on framework/include/bluetooth.h:
```diff
@@ -349,6 +349,10 @@ typedef uint8_t bt_128key_t[16];
 #define BT_UUID_MAX_NUM (32)
 #define BT_UUID_128_LEN (16)
 
+#define BT_INVALID_BROADCAST_ID (0xFFFF)
```
Done.

gzh-terry commented on service/stacks/zephyr/sal_pa_sync_interface.c:
```diff
@@ -0,0 +1,15 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
```
Used in the next PR.

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,220 @@
+static void usage(void)
+{
+    printf("Usage:\n");
+    printf("\taddress: peer device address like 00:01:02:03:04:05\n");
```
Ignored.

gzh-terry commented on framework/socket/bt_pa_sync.c:
```diff
@@ -0,0 +1,17 @@
+ * limitations under the License.
+ ***************************************************************************/
+
+#include "bt_pa_sync.h"
```
Used in the next PR.

gzh-terry commented on Makefile:
```diff
@@ -150,6 +150,11 @@ CSRCS += framework/socket/bt_le_scan.c
+CSRCS += framework/socket/bt_pa_sync.c
+CSRCS += service/ipc/socket/src/bt_socket_pa_sync.c
+endif #CONFIG_BLUETOOTH_PA_SYNC
+
```
Used in the next PR.

gzh-terry commented on service/src/pa_sync_service.c:
```diff
@@ -0,0 +1,67 @@
+
+#include "utils/log.h"
+
+typedef struct pa_sync_info {
+    int nothing_but_wait_for_the_first_variable;
+} pa_sync_info_t;
```
Ignored.

gzh-terry commented on service/src/pa_sync_service.c:
```diff
@@ -0,0 +1,67 @@
+#include "utils/log.h"
+
+typedef struct pa_sync_info {
+    int nothing_but_wait_for_the_first_variable;
```
Ignored.

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,207 @@
+    if (g_bis_sink->scanner) {
+        PRINT("Already scanning");
+        return CMD_USAGE_FAULT;
+    }
```
Done.

chengkai15 commented on service/src/pa_sync_event.h:
```diff
@@ -0,0 +1,74 @@
+#include "bluetooth.h"
+
+/** CTE_Type */
+#define PA_SYNC_EVENT_CTE_TYPE_AOA (0x00) /**< AoA Constant Tone Extension */
```
suggest add prefix for these global macro and enum vars avoid collision.  likes LE_*

gzh-terry replied:
Done.

chengkai15 commented on service/src/pa_sync_event.h:
```diff
@@ -0,0 +1,74 @@
+#define PA_SYNC_EVENT_DATA_TRUNCATED (0x02) /**< Data incomplete, data truncated, no more to come */
+#define PA_SYNC_EVENT_DATA_FAILED (0xFF) /**< Failed to receive an AUX_SYNC_SUBEVENT_IND PDU */
+
+typedef enum {
```
dito

chengkai15 commented on service/stacks/include/sal_pa_sync_interface.h:
```diff
@@ -0,0 +1,94 @@
+    uint8_t cte;
+} bt_sal_pa_sync_param_t;
+
+bt_status_t bt_sal_pa_create_sync(bt_controller_id_t id, const bt_sal_pa_sync_param_t* params);
```
could we replace bt_  prefix with le_?

gzh-terry replied:
I tried to follow the other names, e.g., `bt_sal_gatt_client_enable`, `bt_sal_le_start_scan`

chengkai15 commented on service/stacks/include/sal_pa_sync_interface.h:
```diff
@@ -85,10 +85,69 @@ typedef struct bt_sal_pa_sync_param {
     uint8_t cte;
 } bt_sal_pa_sync_param_t;
 
+/**
```
comment may not need for SAL API,and  keep the style with other modules

gzh-terry replied:
SAL APIs are designed for all 3rd party stacks, so it do need a document.

chengkai15 commented on service/stacks/include/sal_pa_sync_interface.h:
```diff
@@ -85,10 +85,69 @@ typedef struct bt_sal_pa_sync_param {
+ */
 bt_status_t bt_sal_pa_create_sync(bt_controller_id_t id, const bt_sal_pa_sync_param_t* params);
+
+/**
```
dito

gzh-terry replied:
SAL APIs are designed for all 3rd party stacks, so it do need a document.

chengkai15 commented on service/src/pa_sync_service.c:
```diff
@@ -0,0 +1,40 @@
+    BT_LOGD("%s", __func__);
+}
+
+bt_status_t pa_sync_search_terminate(void)
```
could we keep the consistent?
pa_sync_search_delete to  pa_sync_search_create
pa_sync_search_start to pa_sync_search_terminate

gzh-terry replied:
Create and terminate are the official names from core spec.

chengkai15 commented on service/src/pa_sync_service.c:
```diff
@@ -0,0 +1,40 @@
+/****************************************************************************
+ *  Copyright (C) 2025 Xiaomi Corporation
```
2026

gzh-terry replied:
This file started last year :)

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -1917,6 +1920,9 @@ void adapter_init(void)
         BT_LOGE("adapter service state advertise failed :%d", adapter->adapter_state_adv);
 #endif
+#ifdef CONFIG_BLUETOOTH_PA_SYNC
+    pa_sync_init();
```
could we move pa_sync_init to adapter_on_le_enabled?

gzh-terry replied:
`init` and `cleanup` are not related to `enable` and `disable`

Similarly, we do not put `adapter->devices = bt_list_new(adapter_delete_device);` to `adapter_on_le_enabled`

chengkai15 replied:
in my opinion, le feature would not be active until service adapter_on_le_enabled.

so  suggest do it like le adv and gatt does

chengkai15 commented on service/src/adapter_service.c:
```diff
@@ -2027,6 +2033,9 @@ void adapter_cleanup(void)
         orb_unadvertise(adapter->adapter_state_adv);
+#endif
+#ifdef CONFIG_BLUETOOTH_PA_SYNC
+    pa_sync_cleanup();
```
dito

gzh-terry replied:
`init` and `cleanup` are not related to enable and disable

chengkai15 commented on service/src/pa_sync_service.c:
```diff
@@ -17,24 +17,51 @@
+    if (g_pa_sync_info)
+        return BT_STATUS_BUSY;
+
+    g_pa_sync_info = zalloc(sizeof(pa_sync_info_t));
```
could we check g_pa_sync_info  ptr?

gzh-terry replied:
It's checked in the next PR, where `pa_sync_init()` is completed
![Uploading image.png…]()

chengkai15 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -247,12 +247,10 @@ static void adv_data_parsed(adv_data_t* data, void* context)
         return BT_STATUS_FAIL;
 
-    return status;
+    return result->interval ? BT_STATUS_SUCCESS : BT_STATUS_NOT_FOUND;
```
I am some confuse that why check result->interval to decide the parse_adv result？

gzh-terry replied:
Because "have interval" == "have periodic advertising"

chengkai15 replied:
Interval is just one field of PA. It is recommended to check whether PA is effective by checking the content length or flags, etc.

chengkai15 commented on framework/include/bt_le_scan.h:
```diff
@@ -91,11 +91,14 @@ typedef enum {
     bt_address_t addr;
-    uint8_t dev_type; /* bt_device_type_t */
     int8_t rssi;
+    int8_t tx_power; /* transmit power of the advertiser. */
```
it maybe cause crash in cross cores when upgrade in the future

gzh-terry replied:
Should we have a new scan report then?

chengkai15 replied:
agree

zhongzhijie1 replied:
2 core ota , no effect

chengkai15 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -39,9 +39,9 @@ typedef struct {
 
 typedef struct {
-    char value[256];
+    uint8_t value[229]; /**< 229 for extended advertising report */
```
why make the value smaller？

gzh-terry replied:
1. Because the maximum possible adv data is 229 for extended advertising report, the rest room is reserved for extended headers.
2. This struct is removed in the later commits, we can skip this part.

chengkai15 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -85,41 +80,81 @@ static bt_status_t sal_send_req(sal_scan_req_t* req)
-    result_info.adv_type = type;
-    result_info.rssi = rssi;
     result_info.dev_type = BT_DEVICE_DEVTYPE_BLE;
-    result_info.addr_type = addr->type;
-    memcpy(&result_info.addr, &addr->a, sizeof(result_info.addr));
-
-    scan_on_result_data_update(&result_info, eir.value);
+    result_info.adv_type = parse_adv_type(info);
+    result_info.length = ad->len;
+    result_info.rssi = info->rssi;
+    result_info.tx_power = info->tx_power;
+    result_info.sid = info->sid;
+    result_info.interval = info->interval;
+    result_info.addr_type = info->addr->type;
+    memcpy(&result_info.addr, &info->addr->a, sizeof(result_info.addr));
+
+    scan_on_result_data_update(&result_info, ad->data);
 }
 
+static struct bt_le_scan_cb scan_cbs = {
+    .recv = scan_recv_cb,
+    .timeout = NULL,
+};
+
 static void STACK_CALL(start_scan)(void* args)
 {
-    SAL_CHECK(bt_le_scan_start(&scan_param, zblue_on_device_found), 0);
+    int err;
+
+    err = bt_le_scan_cb_register(&scan_cbs);
+    if (err) {
+        BT_LOGE("%s register failed, ret = %d", __func__, err);
+        return;
+    }
+
+    err = bt_le_scan_start(&scan_param, NULL);
+    if (err) {
+        BT_LOGE("%s failed, ret = %d", __func__, err);
+        bt_le_scan_cb_unregister(&scan_cbs);
+    }
```
@gzh-terry please fix it

gzh-terry replied:
Done.

chengkai15 commented on framework/common/advertiser_data_helper.c:
```diff
@@ -240,9 +243,55 @@ bool advertiser_data_parse(const uint8_t* data, uint8_t len, ad_parse_cb_t cb, v
+    switch (uuid_16) {
+#ifdef CONFIG_BLUETOOTH_AURACAST_SINK
+    case BT_UUID_BROADCAST_AUDIO_ANNOUNCEMENT:
+        if (data->len < 1 + sizeof(uuid_16) + 3)
```
could we  make the magic number easy to understand？ l

gzh-terry replied:
Any suggestions?

The name of this uuid is `Broadcast Audio Announcement`

chengkai15 replied:
suggest macro for special size or struct size

chengkai15 commented on tools/bis_sink.c:
```diff
@@ -54,7 +54,18 @@ static void on_scan_result(bt_scanner_t* scanner, ble_scan_result_t* result)
-        return;
+        goto exit;
+
+    if (!result->interval)
```
why check interval?

gzh-terry replied:
If periodic advertising exist:
1. There will be a `SyncInfo` field in the packet.
2. There will be a non-zero `Interval` field in the HCI event.

chengkai15 commented on tools/bis_sink.c:
```diff
@@ -18,6 +18,15 @@
 #include "bt_tools.h"
 
+#define _LOG_STR_LEN (256)
+#define _STRCAT(dst, src, ...)                                         \
```
could expand the access filed ?  and move to utils file

chengkai15 commented on tools/auracast_sink.c:
```diff
@@ -32,11 +32,13 @@ typedef struct {
 static bttool_bis_sink_t* g_bis_sink = NULL;
 static bt_command_t g_bis_sink_tables[] = {
-    { "scan", scan_cmd, 0, "\"Search for nearby Auracast sources\"" },
+    { "scan", scan_start_cmd, 0, "\"Search for nearby Auracast sources\"" },
```
could we make it more frendly? like san start/stop shell tool

gzh-terry replied:
This is to reduce the parameters.
So we can use
`bttool> auracast scan`, rather than
`bttool> auracast scan start`

chengkai15 replied:
Will this tool extend all capabilities of auracast ，including audio、 past etc, or only periodic broadcast?

it seems that it would do somethings likes period adver/scan

gzh-terry committed:
PA sync: add event definitions.

bug: v/82275

The event definition is necessary before the callbacks from periodic advertising sync.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Auracast sink: allow to start extended scan

bug: v/82380

Add scan command to search active Auracast sources.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: add empty service

bug: v/82379

A initial version that has nothing implemented.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: allow to parse scan result

bug: v/82379

Parse the scan result at the client side, so we can reduce the uplink data.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
SAL: add descriptions for PA sync APIs.

bug: v/82275

This patch includes the documentation of SAL APIs.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: allow to parse scan result at SAL

bug: v/82275

Add TxPower, Periodic Advertising Interval, and correct the conversion of adv_type.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
SAL: add definitions for PA sync API.

bug: v/82275

This patch defines the APIs required to synchronize a periodic advertising train.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: API for parsing advertising report

bug: v/82379

Periodic advertising often comes from extended scanning, so we add an parser here.
An alternative is Periodic Advertising Sync Transfer (to be implemented later).

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: parse scan result for bttool

bug: v/82379

Parse scan result at bttol to check if there is an Auracast source available.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: optimize logs at bttool.

bug: v/82379

Skip logs that are not available.
Add more debug logs on error.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: add Kconfig

bug: v/82379

A initial version that starts compiling the files.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: init and uninit for PA sync.

bug: v/82379

Init and uninit for PA sync.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: allow to stop scan

bug: v/82379

Allow to stop scanning for periodic advertisers.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
Auracast sink: Add bttool for Auracast sink

bug: v/82380

Initial version of Auracast sink test tool

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
PA sync: add flags in scan result

bug: v/82820

Add an extra flag to indicate whether periodic advertising is present in the scan result.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry committed:
LE Scan: modify message codes

bug: v/82820

Modify message codes so its safe when client and server have different version of codes.

Signed-off-by: Zihao Gao <gaozihao@xiaomi.com>

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,213 @@
+
+    if (!(result->flags & SCAN_RESULT_FLAG_PERIODIC_ADVERTISING))
+        goto exit;
+
```
Done.

gzh-terry commented on CMakeLists.txt:
```diff
@@ -180,6 +180,11 @@ if(CONFIG_BLUETOOTH)
 
+    if(CONFIG_BLUETOOTH_PA_SYNC)
+      list(APPEND CSRCS ${BLUETOOTH_DIR}/framework/socket/bt_pa_sync.c)
+      list(APPEND CSRCS ${BLUETOOTH_DIR}/service/ipc/socket/src/bt_socket_pa_sync.c)
```
Done

gzh-terry commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +217,92 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+    switch (data->type) {
+    case BT_AD_NAME_SHORT:
+    case BT_AD_NAME_COMPLETE:
+        strlcpy(info->name, (char*)data->data, MIN(sizeof(info->name), data->len));
+        break;
+
+#ifdef CONFIG_BLUETOOTH_AURACAST_SINK
+    case BT_AD_BROADCAST_NAME:
+        strlcpy(info->broadcast_name, (char*)data->data,
+            MIN(sizeof(info->broadcast_name), data->len));
```
The ADV data is a string but without an ending '\0'.

Therefore we need the length `data->len`, which is exactly the length of the string, plus a space for '\0'.

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,213 @@
+
+int auracast_sink_command_init(void* handle)
+{
+    g_auracast_sink = zalloc(sizeof(bttool_auracast_sink_t));
```
Done

gzh-terry commented on tools/utils.h:
```diff
@@ -19,6 +19,14 @@
+        size_t _len = strlen(dst);                               \
+        if (_len + 1 >= (size))                                  \
+            break;                                               \
+        snprintf(dst + _len, (size) - _len, src, ##__VA_ARGS__); \
```
Done

gzh-terry commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -85,41 +80,81 @@ static bt_status_t sal_send_req(sal_scan_req_t* req)
+        return BT_LE_EXT_ADV_IND;
+
+    return BT_LE_EXT_ADV_IND; /**< Unknown */
+};
```
Done

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,213 @@
+    PRINT("%s", log);
+exit:
+    free(info);
+    free(log);
```
Not necessary

gzh-terry commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,213 @@
+    return 0;
+}
+
+int auracast_sink_command_init(void* handle)
+{
+    g_auracast_sink = zalloc(sizeof(bttool_auracast_sink_t));
+
```
Done

gzh-terry commented on framework/common/advertiser_data_helper.c:
```diff
@@ -213,3 +217,92 @@ bool advertiser_data_dump(uint8_t* data, uint16_t len, ad_dump_cb_t dump)
+#ifdef CONFIG_BLUETOOTH_AURACAST_SINK
+    case BT_AD_BROADCAST_NAME:
+        strlcpy(info->broadcast_name, (char*)data->data,
+            MIN(sizeof(info->broadcast_name), data->len));
```
`data->len` does not include the ending '\0', so we do need the orignal `data` to get the full string.

gzh-terry commented on framework/include/bt_le_scan.h:
```diff
@@ -78,12 +80,15 @@ typedef enum {
 typedef struct {
     bt_address_t addr;
+    int8_t rssi; /* RSSI in dBm (-127, +20). 0x7F if unavailable */
+    int8_t tx_power; /* transmit power of the advertiser in dBm (-127, +20). 0x7F if unavailable */
     uint8_t dev_type; /* bt_device_type_t */
-    int8_t rssi;
     uint8_t addr_type; /* ble_addr_type_t */
     uint8_t adv_type; /* ble_adv_type_t */
-    uint8_t length;
-    uint8_t pad[1];
+    uint8_t length; /* length of `adv_data` */
+    uint16_t interval; /* periodic advertising interval in 1.25 microseconds, 0 if not presented */
+    uint8_t sid; /* advertising set identifier, valid from 0x00 to 0x0F, 0xFF if not provided */
+    uint8_t flags; /* e.g., `SCAN_RESULT_FLAG_PERIODIC_ADVERTISING` */
     uint8_t adv_data[1];
 } ble_scan_result_t;
```
Done

gzh-terry commented on framework/include/bt_le_scan.h:
```diff
@@ -78,12 +80,15 @@ typedef enum {
-    uint8_t length;
-    uint8_t pad[1];
+    uint8_t length; /* length of `adv_data` */
+    uint16_t interval; /* periodic advertising interval in 1.25 microseconds, 0 if not presented */
```
Done

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 commented on service/stacks/include/sal_pa_sync_interface.h:
```diff
@@ -85,11 +85,72 @@ typedef struct bt_sal_pa_sync_param {
+ * @brief LE Periodic Advertising Terminate Sync command.
+ *
+ * This command is used to stop reception of the periodic advertising train identified by the
+ * `handle`.
```
What does this mean "by the handle", where is the `handle` definition?

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 commented on tools/auracast_sink.c:
```diff
@@ -36,18 +39,88 @@ static void usage(void)
+        return;
+
+    status = bt_pa_sync_parse_adv_data(info, result);
+    if (status != BT_STATUS_SUCCESS)
```
Free the `info` when return error?

huangyulong3 commented on tools/auracast_sink.c:
```diff
@@ -0,0 +1,65 @@
+static int scan_cmd(void* handle, int argc, char* argv[]);
+
+static bttool_auracast_sink_t* g_auracast_sink = NULL;
+static bt_command_t g_auracast_sink_tables[] = {
```
Add `const` for the table?

gzh-terry replied:
Not possible at this moment.

huangyulong3 reviewed (approved):
LGTM now.

huangyulong3 commented on service/stacks/zephyr/sal_le_scan_interface.c:
```diff
@@ -85,41 +80,81 @@ static bt_status_t sal_send_req(sal_scan_req_t* req)
+    scan_on_result_data_update(&result_info, ad->data);
 }
 
+static struct bt_le_scan_cb scan_cbs = {
```
Add `const` for the callback struct?

gzh-terry replied:
We need to update zephyr api, would do this later.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.

huangyulong3 reviewed (approved):
LGTM.
